ngraph/core/reference/include/ngraph/runtime/reference/convolution.hpp

```diff
@@ -24,338 +25,347 @@
+                    const int filter_size_x = filter.shape[1];
+                    const int dilated_filter_size_x =
+                        filter_size_x + (filter_size_x - 1) * (p.dilation[0] - 1);
+
+                    for (int i_x = -p.pads_begin[0];
+                         i_x <= (p.pads_end[0] + input_size_x - dilated_filter_size_x);
+                         i_x += p.strides[0])
                     {
-                        size_t filter_dilation_stride = filter_dilation[i - 2];
-                        size_t filter_movement_stride = stride[i - 2];
-                        std::ptrdiff_t below_pad = in_pad_below[i - 2];
-                        std::ptrdiff_t above_pad = in_pad_above[i - 2];
-                        size_t in_dilation_stride = in_dilation[i - 2];
-
-                        in_transform_start[i] = filter_movement_stride * out_coord[i];
-                        in_transform_end[i] = in_transform_start[i] +
-                                              (filter_shape[i] - 1) * filter_dilation_stride + 1;
-                        in_transform_movement_strides[i] = filter_dilation_stride;
-                        in_transform_pad_below[i] = below_pad;
-                        in_transform_pad_above[i] = above_pad;
-                        in_transform_dilation_strides[i] = in_dilation_stride;
+                        ACCU sum = 0;
+                        auto input_channel = tensor_slice(batch);
+                        auto filter_channel = tensor_slice(filter);
+                        size_t filter_channels_count = filter.shape[0];
+                        while (filter_channels_count--)
+                        {
+                            for (int f_x = 0; f_x < filter_size_x; ++f_x)
+                            {
+                                int rel_i_x = i_x + (f_x * p.dilation[0]);
+                                bool padding = !in_range(rel_i_x, {0, input_size_x});
+                                if (padding)
+                                    continue;
+
+                                int f_buf_idx = f_x;
+                                int i_buf_idx = rel_i_x;
+
+                                sum += static_cast<ACCU>(input_channel.buffer[i_buf_idx]) *
+                                       static_cast<ACCU>(filter_channel.buffer[f_buf_idx]);
+                            }
+                            ++input_channel;
+                            ++filter_channel;
+                        }
+                        *out.buffer = sum;
+                        ++out.buffer;
                     }
+                }
 
-                    AxisVector in_transform_axis_order(2 + n_spatial_dimensions);
-                    for (size_t i = 0; i < in_transform_axis_order.size(); i++)
+                template <typename INPUT, typename FILTER, typename OUTPUT, typename ACCU>
+                void convolve_2D_channels(const ConvolutionParams& p,
```

Comment:
Could we use the one common implementation for all convolution in order to minimize the potential impact on the binary size?

Reply:
Yes, I will finish today with #3995 and refactor it tomorrow.

Reply:
I removed 1D & 2D implementation in this [commit](https://github.com/openvinotoolkit/openvino/pull/3922/commits/cc79afea07bab9f9342f846e7d0eee4728ecf735) It turned out that I can easily adjust spatial dimensions for 1D & 2D and re-use 3D implementation for all of the cases. 

This implementation is based on assumption that we need to support only up to 3D convolution as in [cpu ref impl](https://github.com/openvinotoolkit/oneDNN/blob/0d8def33381fe359781b036b1da840178973cb0d/src/cpu/ref_convolution.cpp).  If we need to support higher dimensionality that would mean we need to say it explicitly in specification, write tests for such cases and further generalize reference implementation. I believe it is not needed right now, but correct me if I'm wrong. This is important because I will base BinaryConvolution & DeformableConvolution implementations on the same assumptions.